src/component_selector.jl

```diff
@@ -0,0 +1,295 @@
+Optionally provide a name for the `ComponentSelector`.
+"""
+make_selector(content::ComponentSelector...; name::Union{String, Nothing} = nothing) =
+    ListComponentSelector(content, name)
+
+# Naming
+default_name(e::ListComponentSelector) = "[$(join(get_name.(e.content), ", "))]"
+
+# Contents
+function get_groups(e::ListComponentSelector, sys; filterby = nothing)
+    return e.content
+end
+
+function get_components(e::ListComponentSelector, sys::SYSTEM_LIKE; filterby = nothing)
+    sub_components =
+        Iterators.map(x -> get_components(x, sys; filterby = filterby), e.content)
+    return Iterators.flatten(sub_components)
+end
+
+# SubtypeComponentSelector
+"`PluralComponentSelector` represented by a type of component."
+struct SubtypeComponentSelector <: DynamicallyGroupedComponentSelector
+    component_subtype::Type{<:InfrastructureSystemsComponent}
+    name::Union{String, Nothing}
+    groupby::Union{Symbol, Function}
+end
+
+# Construction
+"""
+Make a `ComponentSelector` from a type of component. Optionally provide a name and/or
+grouping behavior for the `ComponentSelector`.
+"""
+make_selector(
+    component_subtype::Type{<:InfrastructureSystemsComponent};
+    name::Union{String, Nothing} = nothing,
+    groupby::Union{Symbol, Function} = :all,
+) =
+    SubtypeComponentSelector(component_subtype, name, validate_groupby(groupby))
+
+# Naming
+default_name(e::SubtypeComponentSelector) = subtype_to_string(e.component_subtype)
+
+# Contents
+function get_components(e::SubtypeComponentSelector, sys::SYSTEM_LIKE; filterby = nothing)
+    components = get_components(e.component_subtype, sys)
+    isnothing(filterby) && (return components)
+    return Iterators.filter(filterby, components)
+end
+
+# FilterComponentSelector
+"`PluralComponentSelector` represented by a filter function and a type of component."
+struct FilterComponentSelector <: DynamicallyGroupedComponentSelector
+    component_subtype::Type{<:InfrastructureSystemsComponent}
+    filter_fn::Function
+    name::Union{String, Nothing}
+    groupby::Union{Symbol, Function}
+end
+
+# Construction
+# Could try to validate filter_fn here, probably not worth it
+"""
+Make a ComponentSelector from a filter function and a type of component. The filter function
+must accept instances of component_subtype as a sole argument and return a `Bool`. Optionally
+provide a name and/or grouping behavior for the `ComponentSelector`.
+"""
+make_selector(
+    component_subtype::Type{<:InfrastructureSystemsComponent},
+    filter_fn::Function; name::Union{String, Nothing} = nothing,
+    groupby::Union{Symbol, Function} = :all,
+) = FilterComponentSelector(component_subtype, filter_fn, name, validate_groupby(groupby))
+
+# Contents
+function get_components(e::FilterComponentSelector, sys::SYSTEM_LIKE; filterby = nothing)
+    components = get_components(e.filter_fn, e.component_subtype, sys)
+    isnothing(filterby) && (return components)
+    return Iterators.filter(filterby, components)
```

Comment:
This makes the function type unstable. We created `FlattenIteratorWrapper` so that we know the length. We shouldn't return different types based on a kwarg.

Reply:
1. Agreed, I'll make these type stable.
2. Looks like not really, not after compilation anyway:
<img width="929" alt="Screenshot 2024-09-30 at 3 20 06 PM" src="https://github.com/user-attachments/assets/447a0b20-45bb-436a-98ce-056d53b78e93">
(this is on the massive EI system)

Reply:
`get_components` now always returns `FlattenIteratorWrapper`. I also ended up combining the two filter functions into one as an implementation detail (not for performance, it just ended up being more convenient with `FlattenIteratorWrapper`).